sqlflow_models/gcn.py

```diff
@@ -0,0 +1,369 @@
+        for i in range(self.nlayer-1):
+            x = tf.keras.activations.relu(self.gc_layers[i](x, adj))
+        x = self.gc_layers[-1](x, adj)
+        return tf.keras.activations.softmax(x)
+
+    @staticmethod
+    def encode_onehot(labels):
+        classes = set(labels)
+        classes_dict = {c: np.identity(len(classes))[i, :] for i, c in enumerate(classes)}
+        labels_onehot = np.array(list(map(classes_dict.get, labels)), dtype=np.int32)
+        return labels_onehot
+
+    @staticmethod
+    def normalize_adj(adjacency, symmetric=True):
+        """
+        Function to normalize the adjacency matrix (get the laplacian matrix).
+        :param adjacency: Adjacency matrix of the dataset.
+            type adjacency: Scipy COO_Matrix.
+        :param symmetric: Boolean variable to determine whether to use symmetric laplacian.
+            type symmetric: bool.
+        """
+        adjacency += sp.eye(adjacency.shape[0])
+        if symmetric:
+            """L=D^-0.5 * (A+I) * D^-0.5"""
+            d = sp.diags(np.power(np.array(adjacency.sum(1)), -0.5).flatten(), 0)
+            a_norm = adjacency.dot(d).transpose().dot(d).tocoo()
+        else:
+            """L=D^-1 * (A+I)"""
+            d = sp.diags(np.power(np.array(adjacency.sum(1)), -1).flatten(), 0)
+            a_norm = d.dot(adjacency).tocoo()
+
+        return a_norm
+
+    @staticmethod
+    def normalize_feature(features):
+        """Function to row-normalize the features input."""
+        rowsum = np.array(features.sum(1))
+        r_inv = np.power(rowsum, -1).flatten()
+        r_inv[np.isinf(r_inv)] = 0.
+        r_mat_inv = sp.diags(r_inv)
+        features = r_mat_inv.dot(features)
+        return features
+
+    @staticmethod
+    def sparse_dropout(x, keep_prob, noise_shape):
+        """Dropout for sparse tensors in Tensorflow."""
+        random_tensor = keep_prob
+        random_tensor += tf.random.uniform(noise_shape)
+        dropout_mask = tf.cast(tf.floor(random_tensor), dtype=tf.bool)
+        pre_out = tf.sparse.retain(x, dropout_mask)
+        return pre_out * (1./keep_prob)
+
+    def preprocess(self, ids, features, labels, edges):
+        """Function to preprocess the node features and adjacency matrix."""
+        if len(features.shape) > 2:
+            features = np.squeeze(features)
+        if len(edges.shape) > 2:
+            edges = np.squeeze(edges)
+        # sort the data in the correct order
+        idx = np.argsort(np.array(ids))
+        features = features[idx]
+        labels = labels[idx]
+        # preprocess
+        features = GCN.normalize_feature(features)
+        labels = GCN.encode_onehot(labels)
+        adjacency = sp.coo_matrix((np.ones(len(edges)),
+                    (edges[:, 0], edges[:, 1])),
+                    shape=(features.shape[0], features.shape[0]), dtype="float32")
+
+        adjacency = adjacency + adjacency.T.multiply(adjacency.T > adjacency) - adjacency.multiply(adjacency.T > adjacency)
```

Comment:
as for 0,1 matrix, the `- adjacency.multiply(adjacency.T > adjacency)` is not necessary.

Reply:
From my perspective, this is still necessary and it aims to build the bidirectional edges of the graph.